src/codegen/intrinsic-applications.lisp

```diff
@@ -0,0 +1,74 @@
+(defpackage #:coalton-impl/codegen/intrinsic-applications
+  (:use #:cl)
+  (:local-nicknames
+   (#:ast #:coalton-impl/codegen/ast)
+   (#:traverse #:coalton-impl/codegen/traverse))
+  (:export
+   #:transform-intrinsic-application
+   #:transform-intrinsic-applications))
+(in-package #:coalton-impl/codegen/intrinsic-applications)
+
+(defgeneric transform-intrinsic-application (rator-name node)
+  (:documentation "Methods for implementing intrinsic-applications to transform
+codegen AST during optimization.
+
+EQL-specialize on symbol `rator-name'.
+`node' will be of type `ast:node-application'.
```

Comment:
It won't be `node-direct-application` because that transform will not have run yet, correct?

Reply:
I think you should just allow both, which you do in your code below.